### PR TITLE
*: replace compareDatum by compare, PR 10

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Related to #29867.

### What is changed and how it works?

This PR handle `compare` related to JOIN.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Small refactoring, more tests will be added later.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: executor/builder.go [hunk 1]
```diff
@@ -2805,23 +2805,29 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plannercore.PhysicalIndexJoin)
 	}
 	innerKeyCols := make([]int, len(v.InnerJoinKeys))
 	innerKeyColIDs := make([]int64, len(v.InnerJoinKeys))
+	keyCtors := make([]collate.Collator, 0, len(v.InnerJoinKeys))
```
Comment (XuHuaiyu) [inline executor/builder.go:2808]: keyCollators

File: executor/builder.go [hunk 2]
```diff
@@ -2805,23 +2805,29 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plannercore.PhysicalIndexJoin)
...
 		innerKeyColIDs[i] = v.InnerJoinKeys[i].ID
+		keyCtors = append(keyCtors, collate.GetCollator(v.InnerJoinKeys[i].RetType.Collate))
 	}
 	e.outerCtx.keyCols = outerKeyCols
 	e.innerCtx.keyCols = innerKeyCols
 	e.innerCtx.keyColIDs = innerKeyColIDs
+	e.innerCtx.keyCollators = keyCtors
 
 	outerHashCols, innerHashCols := make([]int, len(v.OuterHashKeys)), make([]int, len(v.InnerHashKeys))
+	hashCtors := make([]collate.Collator, 0, len(v.InnerHashKeys))
```
Comment (XuHuaiyu) [inline executor/builder.go:2820]: hashCollators

**Review by XuHuaiyu — CHANGES_REQUESTED** [review]

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**